### PR TITLE
feat(core): resolveDependencies — transitive dep pull-through for starter packs (#121)

### LINKED ISSUE #121: System capabilities implementation (Spec 14)
## Summary

Spec 14 defines the system built-in capabilities architecture — LinchKit's own features are delivered as official capability packages, eating its own dog food.

## Split-Out Issues

The following high-priority capabilities have been split into focused issues:

- **cap-audit** → #138 (M5)
- **cap-file-storage** → #139 (M5)
- **cap-notification** → #140 (M7)
- **cap-search** → #141 (M7)

## Remaining Work (this issue)

- **On-demand capabilities:** cap-command-palette, cap-keyboard-shortcuts, cap-theme, cap-report, cap-dashboard, cap-import-export, cap-tag, cap-scheduler, cap-proposal, cap-user-profile
- **Starter packages:** starter-minimal, starter-business, starter-full bundles
- **autoInstall mechanism:** Dependency-based auto-activation for capabilities

## References

- Spec 14: `docs/specs/14_system_capabilities.md`
- Related: Spec 01 (Capability Structure), Spec 20 (Extension Mechanism)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### PR DESCRIPTION
**Pull request by laofahai**
## Summary

- Adds `resolveDependencies(explicit, available)` to `@linchkit/core/server` — the **push-based** complement to the existing `resolveAutoInstall` ("pull") function.
- Exports the new function via `addon-discovery.ts` and updates the barrel surface test.
- Adds 10 new tests covering all edge cases.

## Implementation notes

`resolveAutoInstall` is a **pull** mechanism: "discover me automatically when all my dependencies already appear in the active set." It serves UI/bridge capabilities that auto-activate alongside their backend peers.

`resolveDependencies` is the **push** mechanism: "when I am explicitly installed, also pull in my dependencies from the available pool." This unblocks the starter-pack pattern from Spec 14 §3:

```typescript
// Starter pack usage (Spec 14 §3)
const discovered = await scanAddonsPath(config.addons_path ?? []);
const withDeps   = resolveDependencies(config.capabilities ?? [], discovered);
const activeCaps = resolveAutoInstall(withDeps, discovered);
```

A user can now write:

```typescript
// linchkit.config.ts
capabilities: [capStarterMinimal],   // → resolveDependencies pulls cap-auth + cap-permission
```

instead of listing every transitive dependency explicitly.

**Algorithm**: BFS over `dependencies[]`; fixed-point until stable. Deduplicates by name (explicit version wins over available on collision). Silently skips missing deps — runtime validation surfaces those as hard errors. Circular deps terminate correctly (visited-set guard).

## Spec alignment

- **Spec 14 §3** — Starter package concept (`starter-minimal`, `starter-business`): a starter is a `defineCapability` whose only role is to declare `dependencies`; `resolveDependencies` makes those deps activate automatically.
- **Spec 57 §5** — Discovery flow step 4 ("Merge: explicit capabilities + discovered capabilities"): `resolveDependencies` fills the gap between step 4 and step 5 (autoInstall resolution) by ensuring hard deps of explicit caps are in the active set.

## Quality gates

| Gate | Status |
|------|--------|
| `bun test packages/core/__tests__/capability/` | ✅ 20 pass, 0 fail |
| `bun run check` (whole repo) | ✅ 0 errors, 11 warnings (pre-existing) |
| `bun run typecheck` | ✅ Only pre-existing `bun-types` error (present on `main` too) |
| `bun test packages/core/__tests__/` | ✅ +10 pass vs main (1715 vs 1705), same 103 pre-existing failures |
| `linch validate` | ✅ Not applicable — no meta-model files changed |

## Retro

- **Why this approach:** BFS over `dependencies[]` with a visited set is the standard transitive-closure algorithm — O(V+E), no external deps, easy to test. Explicit wins over available on name collision so users can override dep versions when they know what they're doing.

- **Alternatives considered:**
  1. Modify `resolveAutoInstall` directly — rejected: it would conflate two distinct semantics (pull vs push) and break callers that rely on `autoInstall=true` being the only trigger.
  2. Use `CapabilityHub.resolveDependencyOrder()` — rejected: it works with the `CapabilityManifest` type (Spec 21, ecosystem distribution) not `CapabilityDefinition` (Spec 57, addon runtime). Keeping the two systems separate avoids coupling.
  3. Silently auto-install ALL deps regardless of availability — rejected: YAGNI and could surprise users; missing deps should fail loudly at validation, not silently succeed.

- **Security review:** no auth/tenant/input-trust surfaces touched. This is a pure in-memory algorithm operating on capability metadata arrays. No database access, no HTTP, no user input.

- **Other risks:** Circular dependency graphs do not cause infinite loops (the `active` set prevents re-enqueuing). Missing deps are silently skipped rather than thrown — this is intentional: the validation step (Spec 57 §5 step 7) owns hard-dep errors.

- **Follow-ups:**
  - Wire `resolveDependencies` + `scanAddonsPath` into `cap-adapter-server/src/dev.ts` to fully activate the `addons_path` config flow (currently `addons_path` is configured but never processed at runtime).
  - Create `addons/starter-minimal/cap-starter-minimal/` and `addons/starter-business/cap-starter-business/` definitions that demonstrate the pattern.

## Self-review checklist

- [x] Tests added/updated and passing (10 new, 0 regressions)
- [x] `bun run check` green (0 errors)
- [x] `bun run typecheck` green (pre-existing bun-types error only)
- [x] `bun test` green (+10 vs main, same pre-existing failures)
- [x] `linch validate` green (not applicable — no meta-model files)
- [x] Spec referenced in PR body (Spec 14 §3, Spec 57 §5)
- [x] Mandatory security review walked through — no auth/tenant/input-trust surfaces touched
- [x] No new dependencies
- [x] No hardcoded secrets, no `any`, no `eval`, no `new Function`
- [x] No changes outside the issue's scope (4 files: auto-install.ts, addon-discovery.ts, two test files)
- [x] Branch name follows convention (`feat/resolve-dependencies-121`)
- [x] Commit message follows conventional commits format

Closes #121 (partial — addresses the "autoInstall mechanism: Dependency-based auto-activation" remaining work item)

---
_Generated by [Claude Code](https://claude.ai/code/session_018PWJTQM7YP5ocbHvA6ttRb)_

## Summary by CodeRabbit

* **New Features**
  * Introduced `resolveDependencies` function for computing transitive dependencies in capability requests. Supports circular dependency handling, deduplication, and ensures explicit capabilities take precedence over available ones.

* **Tests**
  * Added comprehensive test suite covering dependency resolution scenarios including transitive resolution, duplicate handling, and edge cases.

[Review Change Stack](https://app.coderabbit.ai/change-stack/laofahai/linchkit/pull/392?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: packages/core/src/capability/auto-install.ts [hunk 1]
```diff
@@ -35,3 +35,65 @@ export function resolveAutoInstall(
...
+      if (active.has(depName)) continue;
+      const dep = byName.get(depName);
+      if (dep) {
+        active.add(depName);
+        result.push(dep);
+        queue.push(dep);
+      }
+      // Missing dep: silently skip; validation will catch it later
+    }
+  }
```
[the remark this answers is not in the document]
Comment (laofahai) [reply packages/core/src/capability/auto-install.ts]: Addressed — the resolution loop was refactored away from `queue.shift()` (which re-indexes the array on every call → O(N²)) to an index-cursor walk over a growing array (`for (let i = 0; i < result.length; i++)` with `result.push(...)`), giving O(N) amortized behavior. Thread is outdated against that refactor.
